drivers/hmis/app/graphql/types/hmis_schema/has_custom_data_elements.rb

```diff
@@ -31,13 +31,22 @@ def custom_data_elements_field(
           define_method(name) do
             resolve_custom_data_elements(object)
           end
+        end
+      end
 
-          define_method(:resolve_custom_data_elements) do |record, definition_scope: nil|
-            # Always resolve all _available_ custom element types for this record type,
-            # (even if they have no value), so that they can be shown as empty if missing.
-            definition_scope ||= Hmis::Hud::CustomDataElementDefinition.for_type(record.class.sti_name)
-            definition_scope
-          end
+      def resolve_custom_data_elements(record)
+        # Load all CustomDataElement values for this record
+        cde_values = load_ar_association(record, :custom_data_elements).group_by(&:data_element_definition_id)
+
+        # Load all CustomDataElementDefinitions that have values for this record
+        cde_definitions = load_ar_association(record, :custom_data_element_definitions)
+
+        # Array of relevant CustomDataElementDefinitions, with value(s)
+        cde_definitions.uniq.map do |cded|
+          OpenStruct.new(
+            **cded.slice(:id, :key, :field_type, :label, :repeats, :show_in_summary),
+            values: cde_values[cded.id],
+          )
```

Comment:
This is the main thing to look at.

I preserved the API schema, just changing how it's loaded. The OpenStruct resolves as a `Types::HmisSchema::CustomDataElement`.
It's slightly annoying that we'll need to keep adding any new attributes on CDEDs into the `slice` on line 47. I support another way this could be set up is just: `OpenStruct.new(definition: cded, values: cde_values[cded.id])`. Any preference there?

Reply:
If we do the second change you suggested, would that be a pretty big change to the api, or would we make a resolution elsewhere so that the type still looks the same from the frontend's perspective?

I wonder if it would be possible/desirable to move the indication of whether the field belongs in the OpenStruct closer to the source, into the CDED model, if that makes sense?-- this is not a very fleshed out idea, but some kind of annotation on the attributes of the CDED model and then here we could just introspect and list through the ones that have that annotation. We might still forget to make the annotation change when adding new fields on that model, but we might be more likely to remember if it's in the model code instead of a faraway concern. But maybe such changes will be infrequent enough to not matter.

Reply:
If we do the second option and it does mean a change to the api, maybe the frontend wants to consume a map instead of an array of cdes? (https://github.com/greenriver/hmis-frontend/pull/680#discussion_r1525062359)

Reply:
Thanks for this great suggestion!

Reply:
I wasn't quite sure where to put the new class. I added it to the same dir as the CustomDataElement models. https://github.com/greenriver/hmis-warehouse/pull/4086/commits/5c26ef44a31939c79f5217e2b80b0742cccbeefd